### DISCUSSION TITLE
Advanced Sonar Equipment

### DISCUSSION THREAD

**Discussion by TheFurNinja** [FakeFishGames/Barotrauma#5219] [Ideas and suggestions]
**Note:** This post is constantly being updated, please check in often and do not be afraid to ask questions! Additionally, note that most values in here such as ranges and angles are placeholders, and are mostly relative to each other for demonstrative effect.

Hey there!

I just started my time in Barotrauma. So far I really love it. But I noticed something immediately: the shallow sonar mechanics.

The sonar itself is beautiful looking and I love how it displays things - but the lack of controls over it (as well as somewhat useless passive) makes it very limited in its capabilities. This short in its capabilities limits the sonar sets ability to expand gameplay. I imagine Europas inhabitants (humans included) are very much tuned into sounds, its a necessary adaptation. A more in depth sonar system could provide more options for players in their gameplay while providing an authentic and immersive submarine experience.

On the thought, I ran to the discord and started bouncing ideas back and fourth between community members to refine a more advanced sonar idea.

Certainly we wanted to make sure that it was intuitive, even for newer players (or even automated, at a cost). Naturally the new system would need to be more capable than the current one. I eventually came up with this (beautiful MS paint) interface (shown below)
![SONAR PASSIVE](https://user-images.githubusercontent.com/32943592/110717011-e8dcdf00-81cd-11eb-96dc-ce70f12d65f1.png)
(displayed here with passive sonar on and scope set to 100m scale)

Additionally we wanted to make sure that the sonar can be used with finesse, so as to give players the option to use the sonar to get detailed information about their surroundings, plan their route, or prepare (assuming trouble is unavoidable). This can allow players to either jump right in and use the system similar to how it functions now, or take their time and potentially avoid threats and/or take advantage of opportunities.

I dont know if this new interface should accompany the Nav Terminal, or just the Sonar Monitor station. It could potentially be an upgrade. I think the devs could use a lot of leeway with my suggestion in general, I just wanna help provide ideas. This was originally intended to be an upgrade or replacement for the Sonar Monitor controls to create a specialized station, but it could be applied elsewhere.

**INTERACTIVE ELEMENTS OF INTERFACE**
Direction and Angle Knobs: The golden knobs on the outside of the circular scope determines the direction and total angle you wish to set the sonar to. Double clicking on the center knob toggles between directional and 360 mode.
ON/OFF Switch: Turns the whole module on/off (we hope).
SCOPE RANGE: 'Zooms' between 300m and 100m scale.
ACTIVE/PASSIVE: Turns sonar set to Active (and emitting) or Passive (listening). Switching to passive switches on the Spectogram scope below.
PING RATE: Switches between HI (One Ping Per Second), Medium (Normal, 1x Ping in 2s ), and LO (1x Ping in 4s). This allows you to be a bit more stealthy (or noisy!) without having to manually fiddle with the passive/active switch.
FREQ: Switches the Frequency of each ping. Each makes audibly different sounds. Each frequency has a different resolution, of which objects below it cannot be detected. Resolution in this case is shown in degrees, which changes the size that an object can be detected at longer ranges. Frequency also affects how well it reflects off of certain materials (almost all objects reflect all frequencies, just not well). An object that reflects too poorly would not appear on the scope.
- HI allows you to see smaller objects and exact shapes using its high (.5°) resolution, but reflects poorly off of large objects at long ranges. It also cannot see small objects at long ranges due to increasing background clutter (false, minimum sized returns - looks like static of increasing size). Noise starts at ~75m and should become an almost unreadable mess at ~200m.
- AUT sweeps across both HI and LO frequencies, making a distinctive sweeping pitched noise (sweeping low to high). Middle ground mode. Has less clutter than HI but also less resolution (2.5°) and doesn't reflect perfectly off of anything (but usually reflects off of most things). 
- LO allows you to see objects clearly from afar, but only large objects like certain monsters or walls due to its very low resolution (5°).

NOISE FILTER: This lever filters the background noise at various levels. Each type of noise has a best setting for clarity
SPECTOGRAM: This scope shows you the signature of all noises being received by the sonar system while it is in passive. The spectogram also puts out an audible sound you can listen to.

**HOW TO OPERATE**
Basic Operations: Flip to on, switch between active/passive at will. Functions similarly to current systems.
Advanced Operations (Active): Flip to on, flip to active, select desired ping rate, set desired frequency, adjust angle and direction knobs to direct the sonar antenna. Active is limited by LOS.
Advanced Operations (Passive): Flip to on, flip to passive, listen for noises, isolate noises using angle and direction knobs, then adjust noise filter until spectrogram signature and noise is clear of 'static'. Passive is not limited by LOS, and is much more sensitive, allowing you to hear through objects (with a large noise penalty) or from great distances in open water. 

**SCENARIOS**
Active:
![SONAR LO-300](https://user-images.githubusercontent.com/32943592/110719876-41fb4180-81d3-11eb-8d28-f949a7299700.png)
Sonar set to Active, LO frequency, and 300m scale - you can see the monster and wall in the distance

![SONAR HI-300](https://user-images.githubusercontent.com/32943592/110720061-90a8db80-81d3-11eb-9a5c-622071e8aad8.png)
Sonar set to Active, HI frequency, and 300m scale - the monster and wall start to reflect poorly (with only parts reflecting), becoming extremely fuzzy and hard to distinguish in the immense noise, but now suddenly you see a small creature up close! (red object)

Passive:
![SONAR ISOLATION](https://user-images.githubusercontent.com/32943592/110720345-00b76180-81d4-11eb-94a7-270225100a2e.png)
Improperly isolated target, spectogram shows a combination of the two signals. You cannot get a perfect clarity on the scope due to the different frequencies involved.

![SONAR FILTER](https://user-images.githubusercontent.com/32943592/110720418-1fb5f380-81d4-11eb-861d-f41b79994ca8.png)
Target is isolated in this case, but the noise filter is improperly set - making the target hard to distinguish.

![SONAR PASSIVE](https://user-images.githubusercontent.com/32943592/110720682-a1a61c80-81d4-11eb-996c-4ae75811adb7.png)
Target is properly isolated and filtered. You can hear it clearly through the hydrophone and can then classify the noise manually.

**INTEGRATION WITH OTHER MODULES**
Nav Terminal: Nav terminal would still retain basic controls. When the Nav terminal is in control it has the gold angle knobs, and can operate the sub in MED rate and AUT freq only @ 100m scale. If the captain desires, he can hand off control of the sonar to the other terminal with the proposed full controls by using another switch. The Captain still retains control of the submarine, but the sonar is then operated by the separate operator.

The Nav terminal also gets access to the audio feed from the passive sonar if they desire by flicking yet another switch, but without access to the noise filter - its capabilities are more limited.

Potential other modules: Auto-Classifier (automatically classifies and tracks noises in passive), Targeting Computer (automatically tracks, classifies, and assists gunnery when sonar is in active).

**End**

That's it for now 😄 

Hopefully this provides a higher skill ceiling and more choices for players while keeping the skill floor equally as low as it is now for operating the sonar. What do others think of my idea?

**Comment by NicolaruS** [comment]
There are custom sonar options built from game logic elements. The use of custom sonar has its limitations, but it can diversify the gameplay.
https://steamcommunity.com/sharedfiles/filedetails/?id=2114129689
https://steamcommunity.com/sharedfiles/filedetails/?id=2114051397

**Reply by TheFurNinja** [reply]
Admittedly these are cool and attempts at implementing something similar to my idea - but having something like my suggestion in the base game would probably be more approachable for players. Especially in base game subs.

**Reply by Krzeszny** [reply]
> There are custom sonar options built from game logic elements. The use of custom sonar has its limitations, but it can diversify the gameplay.
> https://steamcommunity.com/sharedfiles/filedetails/?id=2114129689
> https://steamcommunity.com/sharedfiles/filedetails/?id=2114051397

That looks hella complicated.

**Comment by HeatedWafflez** [comment]
I really like this idea. I understand that making the game friendly to new players is a must but the issue is that if the learning curve is too simple it doesn't really mean much gameplay wise. I think given time people would eventually come to enjoy the more in depth mechanics of a new sonar system because it puts more meaning into being experienced.

**Reply by TheFurNinja** [reply]
Thanks! Thankfully this new interface could technically be used (almost) exactly how ours is now, but stronger in certain situations. This keeps the barrier to entry low. Simultaneously, the full sets of controls would beckon to me, personally. I would spend a good bit of time flipping the switches to see how it behaves. I'm sure on the user end the basics of even the full set of controls would be learned in a few minutes.

Still there would be nuance with reading the scope and knowing how different modes behave. In HI and AUT freq you will have to determine what reflections are clutter (which increases in size as range does) and what isn't. You will eventually know how each frequency reflects off of each material at varying ranges. Comes with some experience

Same with Passive, having it be much stronger and allowing you to actually listen to sounds (potentially through objects if you are very close or it is very loud) gives players an opportunity to use their experience to help the crew. You don't have to, you could just listen to passive similar to how we have it now (just a bit more sensitive) without using the hydrophones.

**Comment by ubertpendragon** [comment]
I like the idea of a better sonar than the one we've got, but that's also lot to take in and I doubt any captain could manage all the tools at the same time

Ignoring the spectrogram, the top part can definitely be simplified with only the addition of a single slider from the existing UI
![image](https://user-images.githubusercontent.com/73229309/111841601-7f13b200-88fe-11eb-965a-aa63fad944b1.png)


ON and OFF are the same as Active/Passive, and if you need a "true" OFF button you can cut the power
RANGE and FREQUENCY can be merged; as you're not going to use high frequency with a 300m scope, nor you'll use low frequency when zoomed in, so it'd adjust automatically. Keep in mind 300m is a massive improvement from the current sonar range
That leaves PING RATE, which would have its own slider

**Reply by TheFurNinja** [reply]
It is a bit buried into the original post, but my primary design was oriented around the Sonar Monitor receiving the new interface, a completely different station. I totally agree that the full controls would be a bit much to manage at the Navigation Terminal while trying to steer the ship and manage the crew, hence suggesting updating the separate station.

I like what you did with the image, though - and it is decently similar to how I described the Navigation Terminal's Proposed Controls.

"Nav Terminal: Nav terminal would still retain basic controls. When the Nav terminal is in control it has the gold angle knobs, and can operate the sub in MED rate and AUT freq only @ 100m scale. If the captain desires, he can hand off control of the sonar to the other terminal with the proposed full controls by using another switch. The Captain still retains control of the submarine, but the sonar is then operated by the separate operator.

The Nav terminal also gets access to the audio feed from the passive sonar if they desire by flicking yet another switch, but without access to the noise filter - its capabilities are more limited."

**Comment by Likvidus** [comment]
I had a similar idea for sonar, you presented it very well. And don't worry about the difficulty for new players, it should be.

**Comment by mygamingaccount** [comment]
I very much like the idea of a long range, listening only (no visual) directional option. Hear something? Is it mechanic like an engine (pirates)?
Of course the background noise would increase the faster you go, so to listen to distant contacts you'd have to slow down.

**Comment by MateusJK15** [comment]
This whole idea is awesome. I always thought that sonar and reactor mechanics are way too simple and boring, realism makes the learning curve wider but also makes the game much more challenging and fun.

**Reply by Krzeszny** [reply]
Right, just look at the popularity of the Neurotrauma mod, the most complex mod to date.
Optional UI elements could be as complex and realistic as possible, while still being relatively easy in auto mode.

**Comment by iridescent-lightning** [comment]
I like this idea. Will try to implement one or two into my mod. Can you join my server so I can better log the feature?

**Comment by Rob-Storm** [comment]
Would really love to see some kind of bearing rate waterfall chart on top of the default sonar display
